# **MÓDULO 30 - AULA 4**
# Implementação em Python


O algoritmo k-means é uma técnica popular de análise de cluster que agrupa dados não rotulados em clusters distintos. Ele é amplamente utilizado em diversas áreas, como análise de mercado, reconhecimento de padrões, e segmentação de clientes. A ideia central por trás do k-means é encontrar k centroides que representem os centros de cada cluster, minimizando a soma dos quadrados das distâncias entre os pontos de dados e os centroides.

Embora o k-means seja eficaz e fácil de implementar, é importante destacar que ele pode ser sensível à inicialização dos centroides e à presença de outliers nos dados. Além disso, a determinação do número ideal de clusters k pode ser um desafio e requer técnicas como o método do cotovelo ou validação interna.

In [ ]:
!pip install plotly --upgrade

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 13.5 MB/s eta 0:00:00
  Attempting uninstall: plotly
    Found existing installation: plotly 5.15.0
    Uninstalling plotly-5.15.0:
      Successfully uninstalled plotly-5.15.0


In [2]:
import plotly.express as px
import plotly.graph_objects as go
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import seaborn as sns

In [3]:
base = pd.read_csv("base_viajantes.csv", delimiter=';')

In [4]:
base

distancia_anual  numero_viagens
0               300               1
1               350               1
2               450               1
3               300               1
4               360               1
5               480               1
6               800               4
7               600               2
8               750               3
9               900               4
10              760               3
11              890               3
12              980               4
13              320               1
14              350               1
15              420               1
16              310               1
17              360               1
18              480               1
19             1000               4
20              900               4
21              870               3
22             1200               5
23              880               3
24              990               4
25             2000               1
26             3200               2
27             3200               1
28             4000               2
29             3500               1
30             3800               1
31             3900               2
32             4500               1
33             4800               2
34             3900               1

In [1]:
sns.pairplot(base)

NameError: name 'sns' is not defined

In [6]:
fig = px.scatter(base, x='distancia_anual', y='numero_viagens',
                 title='Gráfico de Dispersão: Distância Anual vs. Número de Viagens',
                 labels={'distancia_anual': 'Distância Anual', 'numero_viagens': 'Número de Viagens'})

fig.show()

Iremos criar os grupos baseados nos numeros de viagens anuais e distância em km.
Vejam que apenas batendo o olho no gráfico ainda não fica tão claro como estão dividios os grupos.

Para rodar o algoritmo k-means de maneira eficaz, é recomendável padronizar seus dados, especialmente se as variáveis têm escalas diferentes. A padronização (ou normalização) garante que todas as variáveis tenham a mesma escala, evitando que variáveis com magnitudes maiores dominem o processo de clustering.

In [7]:
base_padronizado = base.copy()

In [8]:
colunas_padronizar = ['distancia_anual', 'numero_viagens']
# Inicializar o StandardScaler
scaler = StandardScaler()
# Ajustar e transformar os dados selecionados
base_padronizado[colunas_padronizar] = scaler.fit_transform(base[colunas_padronizar])

In [9]:
base_padronizado

distancia_anual  numero_viagens
0         -0.834483       -0.836599
1         -0.799959       -0.836599
2         -0.730912       -0.836599
3         -0.834483       -0.836599
4         -0.793055       -0.836599
5         -0.710198       -0.836599
6         -0.489248        1.537533
7         -0.627342       -0.045222
8         -0.523771        0.746156
9         -0.420201        1.537533
10        -0.516866        0.746156
11        -0.427105        0.746156
12        -0.364963        1.537533
13        -0.820673       -0.836599
14        -0.799959       -0.836599
15        -0.751626       -0.836599
16        -0.827578       -0.836599
17        -0.793055       -0.836599
18        -0.710198       -0.836599
19        -0.351154        1.537533
20        -0.420201        1.537533
21        -0.440915        0.746156
22        -0.213059        2.328911
23        -0.434010        0.746156
24        -0.358058        1.537533
25         0.339317       -0.836599
26         1.167881       -0.045222
27         1.167881       -0.836599
28         1.720258       -0.045222
29         1.375023       -0.836599
30         1.582164       -0.836599
31         1.651211       -0.045222
32         2.065493       -0.836599
33         2.272634       -0.045222
34         1.651211       -0.836599

**Hora de aplicarmos o algoritmo do k means**

In [10]:
kmeans_perfil_viajante = KMeans (n_clusters=2, n_init=10, random_state=42)
kmeans_perfil_viajante.fit(base_padronizado)

KMeans(n_clusters=2, n_init=10, random_state=42)

n_clusters=3: Este parâmetro define o número de clusters que você deseja identificar nos seus dados. No exemplo dado, n_clusters=3 significa que você está pedindo ao algoritmo k-means para agrupar seus dados em três clusters distintos.

n_init=10: Este parâmetro define o número de vezes que o algoritmo k-means será executado com diferentes centroides iniciais.

random_state=42: Este parâmetro é usado para controlar a aleatoriedade na inicialização dos centroides. Definir um random_state garante que os resultados sejam reproduzíveis entre diferentes execuções do código.

In [11]:
centroides = kmeans_perfil_viajante.cluster_centers_
centroides

array([[-0.59972289,  0.20801919],
       [ 1.49930722, -0.52004798]])

In [12]:
centroides_padronizados = kmeans_perfil_viajante.cluster_centers_  # Centroides padronizados retornados pelo k-means

# Inverter a padronização para obter os centroides originais
centroides_originais = scaler.inverse_transform(centroides_padronizados)
centroides_originais

array([[6.40e+02, 2.32e+00],
       [3.68e+03, 1.40e+00]])

O primeiro centroide [3680, 1.40] sugere que, no cluster representado por esse centroide, os dados têm uma média de aproximadamente 3680 unidades para a variável distancia_anual e uma média de cerca de 1.40 para a variável numero_viagens.

O segundo centroide [640, 2.32] indica que no cluster associado a este centroide, os dados têm uma média de cerca de 640 unidades para distancia_anual e uma média de aproximadamente 2.32 para numero_viagens.

Agora criando os labels dos grupos:

In [13]:
labels = kmeans_perfil_viajante.labels_
labels

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [14]:
# Criando um df com os dados originais e  rótulos de cluster
df_clusters = pd.DataFrame({
    'distancia_anual': base['distancia_anual'],
    'numero_viagens': base['numero_viagens'],
    'cluster': labels.astype(str)
})

# Plotar os pontos de dados coloridos
fig = px.scatter(df_clusters, x='distancia_anual', y='numero_viagens', color='cluster',
                 color_continuous_scale='viridis', opacity=0.7,
                 title='Clusters de Perfil de Viajantes')

# Adicionar os centroides ao gráfico
fig.add_scatter(x=centroides_originais[:, 0], y=centroides_originais[:, 1],
                mode='markers', marker=dict(color='red', symbol='x', size=14),
                name='Centroides')


fig.update_layout(xaxis_title='Distância Anual', yaxis_title='Número de Viagens',
                  legend_title='Cluster')
fig.show()

Vemos que o centroide 1 do clusterr 1 parece estar bem distantes dos pontos daqueles que viajam mais que 3 vezes. Vamos tentar rodar o modelo novamente definindo 3 clusters.

In [15]:
kmeans_perfil_viajante_3 = KMeans (n_clusters=3, n_init=10, random_state=42)
kmeans_perfil_viajante_3.fit(base_padronizado)

KMeans(n_clusters=3, n_init=10, random_state=42)

In [16]:
centroides_padronizado_3 = kmeans_perfil_viajante_3.cluster_centers_
centroides_originais_3  = scaler.inverse_transform(centroides_padronizado_3)
centroides_originais_3

array([[5.05714286e+02, 1.07142857e+00],
       [3.86666667e+03, 1.44444444e+00],
       [9.10000000e+02, 3.66666667e+00]])

In [17]:
labels_3 = kmeans_perfil_viajante_3.labels_
labels_3

array([0, 0, 0, 0, 0, 0, 2, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 2, 2, 2,
       2, 2, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [18]:
df_clusters = pd.DataFrame({
    'distancia_anual': base['distancia_anual'],
    'numero_viagens': base['numero_viagens'],
    'cluster': labels_3.astype(str)
})

fig = px.scatter(df_clusters, x='distancia_anual', y='numero_viagens', color='cluster',
                 color_continuous_scale='viridis', opacity=0.7,
                 title='Clusters de Perfil de Viajantes')

fig.add_scatter(x=centroides_originais_3[:, 0], y=centroides_originais_3[:, 1],
                mode='markers', marker=dict(color='red', symbol='x', size=14),
                name='Centroides')


fig.update_layout(xaxis_title='Distância Anual', yaxis_title='Número de Viagens',
                  legend_title='Cluster')
fig.show()

In [19]:
base['cluster'] = labels_3

In [20]:
medias_por_cluster = base.groupby('cluster')[['distancia_anual', 'numero_viagens']].mean()

In [21]:
medias_por_cluster


distancia_anual  numero_viagens
cluster                                 
0             505.714286        1.071429
1            3866.666667        1.444444
2             910.000000        3.666667

Conseguimos identificar 3 grupos de segmentação de clientes:

0 - Viaja uma distância média de 3866 km por ano e em torno de 1 a 2 viagens. = Viagens mais longas, menos vezes por ano. Podemos inferir que viagens mais caras, viagens internacionais.

2 - Viaja uma distância média de 505 km e em torno 1 viagem por ano.
= Viagens mais curtas, nacionais, viagem anual, mais baratas.

3 - 910 km e em torno de 3 a 4 viagens, são viagens curtas e várias vezes no ano. Podemos inferir que esse grupo viaja a trabalho.

Através dessa análise poderiamos dar informações valiosas para o time de marketing por exemplo. Oferecer as viagens mais longas via email para os clientes do grupo 0, e verificar se eles já viajaram no ano, se já viajaram 2x é provavel que não viagem mais. Para os clientes no grupo 2 poderiamos enviar viagens promocionais nacionais.
Já os clientes do grupo 2 valeria entender os destinos que viajam a trabalho e a recorrencia e enviar emails de lembretes com ofertas.